salt/_modules/metalk8s_drain.py

```
@@ -0,0 +1,512 @@
+            api_response = api_instance.list_pod_for_all_namespaces(
+                field_selector='spec.nodeName={0}'.format(self._node_name)
+            )
+        except (ApiException, HTTPError) as exc:
+            if isinstance(exc, ApiException) and exc.status == 404:
+                return []
+            log.exception(
+                'Exception when calling '
+                'CoreV1Api->list_pod_for_all_namespaces')
+            raise CommandExecutionError(exc)
+
+        for pod in api_response.items:
+            pod_deletable = True
+            for filt in (
+                    _mirrorpod_filter,
+                    self.localstorage_filter,
+                    self.unreplicated_filter,
+                    self.daemonset_filter
+            ):
+                try:
+                    filter_deletable, warning = filt(pod)
+                except DrainException as exc:
+                    failures.setdefault(
+                        exc.message, []).append(pod.metadata.name)
+
+                if warning:
+                    warnings.setdefault(
+                        warning, []).append(pod.metadata.name)
+                pod_deletable = pod_deletable and filter_deletable
+            if pod_deletable:
+                pods.append(pod)
+
+        def message(errors):
+            msgs = []
+            for key, msg in errors.items():
+                msgs.append("{0}: {1}".format(key, ", ".join(msg)))
+            return "; ".join(msgs)
+
+        if failures:
+            raise DrainException(message(failures))
+        if warnings:
+            log.warning("WARNING: %s", message(warnings))
+        return pods
+
+    def run_drain(self, dry_run=False, **kwargs):
+        '''Cordon the node if necessary, then drain it.'''
+        node_unschedulable = __salt__['metalk8s_kubernetes.node_unschedulable'](
+            self.node_name, **kwargs
+        )
+        if not node_unschedulable:
```

TOCTOU risk here: why not try to cordon and ignore any error if it's already done?

---

Indeed. Cordon is idempotent anyway.